### PR TITLE
produce progress events polling ctrd's content.Store

### PR DESCRIPTION
**Pull request by ndeloof**
**- What I did**
render `docker pull` progress

**- How I did it**
poll `content.Store` for active pulls and push`Progress` data according to moby's API

### CODE REVIEW

**Review by vvoland — APPROVED** [review]

**Review by rumpl — APPROVED** [review]

File: daemon/containerd/service.go [hunk 1]
```diff
@@ -75,7 +75,38 @@ func (cs *containerdStore) PullImage(ctx context.Context, image, tag string, pla
...
+		showProgress(ctx, jobs, cs.client.ContentStore(), outStream, stop)
+		stop <- struct{}{}
+	}()
+
+	img, err := cs.client.Pull(ctx, ref.String(), opts...)
+	if err != nil {
+		return err
+	}
+
+	unpacked, err := img.IsUnpacked(ctx, "overlayfs")
```
Comment (rumpl) [inline daemon/containerd/service.go:98]: Maybe we should use `containerd.DefaultSnapshotter` here

File: daemon/containerd/service.go [hunk 2]
```diff
@@ -75,7 +75,38 @@ func (cs *containerdStore) PullImage(ctx context.Context, image, tag string, pla
...
+	if err != nil {
+		return err
+	}
+
+	if !unpacked {
+		if err := img.Unpack(ctx, "overlayfs"); err != nil {
```
Comment (rumpl) [inline daemon/containerd/service.go:104]: Same comment as above

**Comment by thaJeztah** [comment]
When upstreaming:

- depends on / needs to go after https://github.com/moby/moby/pull/43819
- but, we may want to squash all fixes related to "common" progress (push/pull) code into a single commit

combine with one or more of;

- https://github.com/rumpl/moby/pull/15
- https://github.com/rumpl/moby/pull/16
- https://github.com/rumpl/moby/pull/73
- https://github.com/rumpl/moby/pull/74
- progress fix-up from https://github.com/rumpl/moby/pull/80

File: daemon/containerd/service.go [hunk 3]
```diff
@@ -75,7 +75,38 @@ func (cs *containerdStore) PullImage(ctx context.Context, image, tag string, pla
...
+		return err
+	}
+
+	if !unpacked {
+		if err := img.Unpack(ctx, containerd.DefaultSnapshotter); err != nil {
+			return err
+		}
+	}
+	stop <- struct{}{}
+	<-stop
```
Comment (thaJeztah) [inline daemon/containerd/service.go:109]: Heads up: this code is removed in a later PR, so (perhaps?) can be squashed

- https://github.com/rumpl/moby/pull/64

**Pull request merged**